File: tools/parity/dispatch/policy.ts
// TS dispatch for the policy parity module: drives the LIVE
// src/main/story-world/play-path-guard.ts and src/shared/fleet-grant.ts against
// the Rust port (orca-policy).
//
// These two are the fleet's authority decisions — "may this file be served to a
// child's browser", "may this caller type into that agent's terminal" — so the
// differential harness matters more here than for a formatter: a divergence is a
// security divergence, and the corpus already caught one (the two sides
// disagreed about WHERE percent-decoding happens).
//
// Only the LEXICAL half of the play-path decision is compared. The realpath half
// needs a filesystem, so `realpath` is stubbed to identity and containment
// against a symlinked world is covered by the guard's own suite.

import { relative, sep } from 'node:path'
import { decidePlayPath } from '../../../src/main/story-world/play-path-guard'
import { isAllowedPlayHost } from '../../../src/main/story-world/play-path-guard'
import {
  decideFleetGrant,
  type FleetGrant,
  type FleetGrantOp
} from '../../../src/shared/fleet-grant'

/** Any absolute root works; the compared output is the path RELATIVE to it. */
const ROOT = '/r'

type GrantVector = {
  generation?: number
  ops?: string[]
  targets?: { handle?: string; incarnation?: string | null }[]
  expiresAtMs?: number | null
  revoked?: boolean
}

/**
 * Rebuilds the vector's grant into the live TS shape. `grantId`/`runId`/`secret`
 * are echoes, not inputs to the decision, so they are filled with constants and
 * dropped from the comparison below.
 */
function toFleetGrant(raw: GrantVector): FleetGrant {
  return {
    grantId: 'g',
    runId: 'r',
    secret: 's',
    generation: raw.generation ?? 0,
    ops: (raw.ops ?? []) as readonly FleetGrantOp[],
    targets: (raw.targets ?? []).map((target) => ({
      handle: target.handle ?? '',
      incarnationId: target.incarnation ?? null
    })),
    expiresAt: raw.expiresAtMs ?? null,
    // The Rust core carries a boolean; the TS grant carries the revocation
    // TIMESTAMP, and only its null-ness is read. 1 is any non-null instant.
    revokedAt: raw.revoked === true ? 1 : null
  }
}

export function dispatch(fn: string, input: unknown): unknown {
  const args = (input ?? {}) as Record<string, unknown>
  switch (fn) {
    case 'decidePlayPathLexical': {
      const decision = decidePlayPath({
        root: ROOT,
        requestPath: String(args.requestPath ?? ''),
        realpath: (path) => path
      })
      if (!decision.allowed) {
        return { allowed: false, reason: decision.reason }
      }
      // Separators normalized to `/` so the compared value is the same on
      // Windows, where `resolve`/`relative` return backslashes.
      return {
        allowed: true,
        relativePath: relative(ROOT, decision.absolutePath).split(sep).join('/')
      }
    }
    case 'isAllowedPlayHost':
      return isAllowedPlayHost(
        args.host === undefined ? undefined : String(args.host),
        Number(args.expectedPort ?? 0)
      )
    case 'decideFleetGrant': {
      const grant =
        args.grant === null || args.grant === undefined
          ? null
          : toFleetGrant(args.grant as GrantVector)
      const decision = decideFleetGrant(grant, {
        op: String(args.op ?? '') as FleetGrantOp,
        handle: String(args.handle ?? ''),
        incarnationId: args.incarnation === undefined ? null : (args.incarnation as string | null),
        generation: Number(args.currentGeneration ?? 0),
        now: Number(args.nowMs ?? 0)
      })
      // Projected to the decision itself. `grantId`/`runId` are echoes of the
      // input that the Rust core deliberately does not carry.
      return decision.allowed ? { allowed: true } : { allowed: false, reason: decision.reason }
    }
    default:
      return { __parity_error__: `unknown function ${fn}` }
  }
}

File: rust/crates/orca-dispatch/src/modules/policy.rs
//! Parity dispatch for `orca_policy` vs its TS twins
//! `src/main/story-world/play-path-guard.ts` and `src/shared/fleet-grant.ts`.
//!
//! Unlike most modules here, these two are *authority* decisions rather than
//! formatting or ranking: "may this file be served to a child's browser" and
//! "may this caller type into that agent's terminal". The parity seam exists so
//! the verified Rust answer and the shipping TS answer cannot drift apart
//! silently — the shared `parity-corpus.txt` is run by both sides.
//!
//! Only the LEXICAL half of the play-path decision crosses this seam. The
//! realpath half needs a filesystem, so it stays where the syscall is.

use orca_policy::{
    decide_fleet_grant, decide_play_path_lexical, is_allowed_play_host, Grant, GrantRequest,
    GrantTarget, PlayPathVerdict,
};
use serde_json::{json, Value};

fn string_at(input: &Value, key: &str) -> String {
    input
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

/// An absent/non-string field reads as `None`, which is the FAIL-CLOSED side for
/// `incarnation`: unknown incarnation means a respawn cannot be ruled out.
fn optional_string_at(input: &Value, key: &str) -> Option<String> {
    input
        .get(key)
        .and_then(Value::as_str)
        .map(ToString::to_string)
}

fn string_list_at(input: &Value, key: &str) -> Vec<String> {
    input
        .get(key)
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Value::as_str)
                .map(ToString::to_string)
                .collect()
        })
        .unwrap_or_default()
}

fn parse_grant(value: &Value) -> Grant {
    Grant {
        generation: value.get("generation").and_then(Value::as_u64).unwrap_or(0),
        ops: string_list_at(value, "ops"),
        targets: value
            .get("targets")
            .and_then(Value::as_array)
            .map(|items| {
                items
                    .iter()
                    .map(|target| GrantTarget {
                        handle: string_at(target, "handle"),
                        incarnation: optional_string_at(target, "incarnation"),
                    })
                    .collect()
            })
            .unwrap_or_default(),
        expires_at_ms: value.get("expiresAtMs").and_then(Value::as_u64),
        revoked: value
            .get("revoked")
            .and_then(Value::as_bool)
            .unwrap_or(false),
    }
}

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        // Returns the TS `PlayPathDecision` shape minus `absolutePath`, which
        // only the realpath half can supply.
        "decidePlayPathLexical" => match decide_play_path_lexical(&string_at(input, "requestPath"))
        {
            PlayPathVerdict::NeedsRealpathCheck { relative_path } => {
                json!({ "allowed": true, "relativePath": relative_path })
            }
            PlayPathVerdict::Denied(denial) => {
                json!({ "allowed": false, "reason": denial.as_str() })
            }
        },
        // An absent Host header reads as the empty string, which has no port and
        // is therefore refused — same as the TS `if (!host) return false`.
        "isAllowedPlayHost" => json!(is_allowed_play_host(
            input.get("host").and_then(Value::as_str).unwrap_or_default(),
            u16::try_from(
                input
                    .get("expectedPort")
                    .and_then(Value::as_u64)
                    .unwrap_or(0)
            )
            .unwrap_or(0),
        )),
        "decideFleetGrant" => {
            let grant = input.get("grant").filter(|g| !g.is_null()).map(parse_grant);
            let request = GrantRequest {
                op: string_at(input, "op"),
                handle: string_at(input, "handle"),
                incarnation: optional_string_at(input, "incarnation"),
                current_generation: input
                    .get("currentGeneration")
                    .and_then(Value::as_u64)
                    .unwrap_or(0),
                now_ms: input.get("nowMs").and_then(Value::as_u64).unwrap_or(0),
            };
            match decide_fleet_grant(grant.as_ref(), &request) {
                None => json!({ "allowed": true }),
                Some(denial) => json!({ "allowed": false, "reason": denial.as_str() }),
            }
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
